File: app/src/main/java/de/bikebean/app/ui/utils/permissions/PermissionsRationaleDialog.java
package de.bikebean.app.ui.utils.permissions;

import android.app.Activity;
import android.app.Dialog;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

import java.util.HashMap;
import java.util.Map;

import de.bikebean.app.R;

public class PermissionsRationaleDialog extends DialogFragment {

    private final @NonNull Activity act;
    private final @NonNull PermissionUtils.KEYS permissionKey;

    private static final Map<PermissionUtils.KEYS, Integer[]> permissionMap =
            new HashMap<PermissionUtils.KEYS, Integer[]>() {{
                put(PermissionUtils.KEYS.SMS, new Integer[]{
                        R.string.no_sms_permission,
                        R.string.no_sms_permission_warning
                });
                put(PermissionUtils.KEYS.MAPS, new Integer[]{
                        R.string.no_map_permission,
                        R.string.no_map_permission_warning
                });
    }};

    public PermissionsRationaleDialog(final @NonNull Activity act,
                                      final @NonNull PermissionUtils.KEYS permissionKey) {
        this.act = act;
        this.permissionKey = permissionKey;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
        final @NonNull AlertDialog.Builder builder = new AlertDialog.Builder(act);

        final @Nullable Integer[] integers = permissionMap.get(permissionKey);
        final @Nullable Integer title;
        final @Nullable Integer message;
        if (integers != null) {
            title = integers[0];
            message = integers[1];
        } else {
            title = null;
            message = null;
        }

        if (title != null && message != null)
            builder.setTitle(title)
                    .setMessage(message);
        else
            builder.setTitle("Berechtigungen")
                    .setMessage("Die App benötigt weitere Berechtigungen!");

        builder.setPositiveButton(R.string.continue_ok, (dialog, id) ->
                PermissionUtils.askForPermissions(act, permissionKey)
        );

        return builder.create();
    }
}

File: app/src/main/java/de/bikebean/app/ui/utils/permissions/PermissionsRationaleDialog.kt
package de.bikebean.app.ui.utils.permissions

import android.app.Activity
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.Fragment
import de.bikebean.app.R
import de.bikebean.app.ui.utils.permissions.PermissionUtils.KEYS
import de.bikebean.app.ui.utils.permissions.PermissionUtils.askForPermissions

class PermissionsRationaleDialog(
        private val permissionKey: KEYS,
        private val _context: Context,
        private val act: Activity? = null,
        private val fragment: Fragment? = null,
) : DialogFragment() {

    constructor(act: Activity, permissionKey: KEYS) : this(
            permissionKey, act, act
    )

    constructor(fragment: Fragment, permissionKey: KEYS) : this(
            permissionKey, fragment.requireContext(), fragment = fragment
    )

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val (title, message) = permissionMap[permissionKey] ?: default

        val buttonClickListener: ((DialogInterface, Int) -> Unit)? = when {
            act != null -> { _, _ -> askForPermissions(act, permissionKey) }
            fragment != null ->  { _, _ -> askForPermissions(fragment, permissionKey) }
            else -> null
        }

        return AlertDialog.Builder(_context).apply {
            setTitle(title)
            setMessage(message)
            setPositiveButton(R.string.button_ok, buttonClickListener)
        }.create()
    }

    companion object {
        private val permissionMap = mapOf(
                KEYS.SMS to (R.string.title_permission_sms to R.string.message_permission_sms),
                KEYS.MAPS to (R.string.title_permissions_map to R.string.message_permission_map)
        )

        private val default = R.string.title_permission_generic to R.string.message_permission_generic
    }
}